substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/PreMainSupport.java

```diff
@@ -0,0 +1,273 @@
+            throw new UnsupportedOperationException("Native image doesn't support retransform class at runtime.");
+        }
+
+        @Override
+        public boolean isRedefineClassesSupported() {
+            return false;
+        }
+
+        @Override
+        public void redefineClasses(ClassDefinition... definitions) throws ClassNotFoundException, UnmodifiableClassException {
+            throw new UnsupportedOperationException("Native image doesn't support redefine class at runtime.");
+        }
+
+        @Override
+        public boolean isModifiableClass(Class<?> theClass) {
+            return false;
+        }
+
+        @Override
+        public Class<?>[] getAllLoadedClasses() {
+            synchronized (this) {
+                if (allLoadedClasses == null) {
+                    allLoadedClasses = ClassForNameSupport.singleton().getAllLoadedClasses();
+                }
+            }
+            return allLoadedClasses;
+        }
+
+        @Override
+        public Class<?>[] getInitiatedClasses(ClassLoader loader) {
+            throw new UnsupportedOperationException("Native image has flat classloader hierarchy. Can't get classes by classloader. " +
+                            "Try to call getAllLoadedClasses() to get all loaded classes.");
+        }
+
+        @Override
+        public long getObjectSize(Object objectToSize) {
+            return -1;
+        }
+
+        @Override
+        public void appendToBootstrapClassLoaderSearch(JarFile jarfile) {
+        }
+
+        @Override
+        public void appendToSystemClassLoaderSearch(JarFile jarfile) {
+        }
+
+        @Override
+        public boolean isNativeMethodPrefixSupported() {
+            return false;
+        }
+
+        @Override
+        public void setNativeMethodPrefix(ClassFileTransformer transformer, String prefix) {
+        }
+
+        @Override
+        public void redefineModule(Module module, Set<Module> extraReads, Map<String, Set<Module>> extraExports, Map<String, Set<Module>> extraOpens, Set<Class<?>> extraUses,
+                        Map<Class<?>, List<Class<?>>> extraProvides) {
+            if (!module.isNamed()) {
+                return;
+            }
+
+            if (!isModifiableModule(module)) {
+                throw new UnmodifiableModuleException(module.getName());
+            }
+
+            for (Module extraRead : extraReads) {
+                ModuleNative.addReads(module, extraRead);
+            }
+
+            for (Map.Entry<String, Set<Module>> entry : extraExports.entrySet()) {
+                for (Module m : entry.getValue()) {
+                    ModuleNative.addExports(module, entry.getKey(), m);
+                }
+            }
+            // Ignore the extraOpens, extraUses and extraProvides as they are not supported by
+            // ModuleNative.
+        }
+
+        @Override
+        public boolean isModifiableModule(Module module) {
```

Comment:
@ivan-ristovic can you please look at this implementation?

Reply:
LGTM. This is the same implementation as in `java.instrument` API, and our run-time module system has no specific quirks that require divergence from the JDK.